### PR TITLE
Style Guide: Functional vs object-oriented style, #22805

### PR DESCRIPTION
**Pull request by patriknw**
First section of the style guide, covering function vs object-oriented style.

Refs #22805

### CODE REVIEW

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+* There is no class in the functional style, but that is not strictly a requirement and sometimes it's
+  convenient to use a class also with the functional style to reduce number of parameters in the methods.
+* Mutable state, such as the @scala[`var n`]@java[`int n`] is typically used in the object-oriented style.
+* In the functional style the state is is updated by returning a new behavior that holds the new immutable state,
+  the @scala[`n: Int`]@java[`final int n`] parameter of the `counter` method.
+* The object-oriented style must use a new instance of the initial `Behavior` for each spawned actor instance,
+  since the state in `AbstractBehavior` instance must not be shared between actor instances.
+  This is "hidden" in the functional style since the immutable state is captured by the function.
+* In the object-oriented style one can return `this` to stay with the same behavior for next message.
+  In the functional style there is no `this` so `Behviors.same` is used instead.
```
Comment (raboof) [inline akka-docs/src/main/paradox/typed/style-guide.md]:
```suggestion
  In the functional style there is no `this` so `Behaviors.same` is used instead.
```

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+* It could be more familiar and easier to migrate existing classic actors to this style.
+* Mutable state can sometimes have better performance, e.g. mutable collections and
+  avoiding allocating new instance for next behavior (be sure to benchmark if this is your
+  motivation).
+
+@@@
+
+@@@ div {.group-java}
+
+When developing in Java the functional style will probably be the choice for many.
```
Comment (raboof) [inline akka-docs/src/main/paradox/typed/style-guide.md]:
```suggestion
When developing in Java the object-oriented style will probably be the choice for many.
```
I think?

**Review by raboof — APPROVED** [review]
Good guidance! 2 comments.

File: akka-docs/src/main/paradox/typed/index.md [hunk 3]
```diff
@@ -22,5 +22,6 @@
 * [persistence](persistence.md)
 * [fsm](fsm.md)
 * [testing](testing.md)
+* [actors](style-guide.md)
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/index.md]: s/actors/style/

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+When developing in Scala the functional style will probably be the choice for many.
+
+Some reasons why you may want to use the functional style:
+
+* You are familiar with a functional approach of structuring the code. Note that this API is still
+  not using any advanced functional programming or type theory constructs.
+* The state is immutable and can be passed to "next" behavior.
+* The `Behavior` is stateless.
+* The actor lifecycle has several different phases that can be represented by switching between different
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/style-guide.md]: link to behaviors as FSMs?

**Review by chbatey — APPROVED** [review]
LGTM, Maybe explicitly mention some where that neither is more powerful

File: akka-docs/src/main/paradox/typed/style-guide.md (hunk 2 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/typed/style-guide.md]: thanks

**Pull request merged**